### PR TITLE
Balance tweaks, QOL & some cleanup

### PR DESCRIPTION
**Pull request by StrawWagen**
All panels now close when you click off of them, press use again, or open the Q menu.

Turrets have been split into two categories, heavy turrets and light turrets
Based off testing, gives a use to the crap turrets, while limiting the super strong ones.

Light turrets (smg, pistol, plinker)
spawn with ammo & a battery already attached. 
Max 4 light turrets

Heavy turrets( tesla, grenade, machinegun, shotgun, sniper )
They're how turrets were before, you have to manually attach batteries/ammo 
They now drain battery a lot faster ( hook them up to a generator! ) 
Max 2 heavy turrets.

NPCS now target/fight the turrets.

Grenade turret now has a proper tracer, can actually figure out where you're being shot from.

Made generator turning on/off less vague

Spam the "broken" effect less ( affects implanter, teslaturret, normal turret )

Paintcan now defaults to last applied color.

Can now swap out batteries when they're below 25% charge ( had to be completely empty before ) 
Same with ammo. ( has to be below five rounds )

Grenade turret damage, firerate buff.
Range, HP, max ammo, grenade flight speed nerf.

Sniper turret range, firerate, damage buff.
Small HP nerf

Barbed wire no longer takes ACF damage.

### CODE REVIEW

File: materials/entities/ent_jack_aidcomm.png
Comment (wrefgtzweve) [inline materials/entities/ent_jack_aidcomm.png:1]: This file can be removed alongside other pngs that dont exist for our current entities

File: lua/autorun/client/jid_client.lua [hunk 1]
```diff
@@ -16,3 +16,43 @@ killicon.Add( "ent_jack_seamine", "vgui/hud/jid_seamine_killicon", iconColor )
 killicon.Add( "ent_jack_landmine", "vgui/hud/jid_landmine_killicon", iconColor )
 killicon.Add( "ent_jack_boundingmine", "vgui/hud/jid_boundingmine_killicon", iconColor )
 killicon.Add( "ent_jack_teslasentry", "vgui/hud/jid_teslasentry_killicon", iconColor )
+
+local input_IsKeyDown = input.IsKeyDown
```
Comment (wrefgtzweve) [inline lua/autorun/client/jid_client.lua]: Localizing these seems rather pointless

File: lua/effects/eff_jack_rocketthrust/init.lua [hunk 2]
```diff
@@ -1,3 +1,6 @@
+local math = math
```
Comment (wrefgtzweve) [inline lua/effects/eff_jack_rocketthrust/init.lua]: I dont see the point of localizing `math` here, same for `VectorRand` and in other files.

File: lua/entities/ent_jack_40mmgrenade/init.lua [hunk 3]
```diff
@@ -72,3 +110,16 @@ function ENT:Detonate()
 
     self:Remove()
 end
+
+function ENT:OnRemove()
+    local smokeTrail = self.SmokeTrail
+    if IsValid( smokeTrail ) then
+        SafeRemoveEntityDelayed( smokeTrail, smokeTrailLifetime )
+
```
Comment (wrefgtzweve) [inline lua/entities/ent_jack_40mmgrenade/init.lua]: Empty newline should be removed and added after the `end` instead, same for the tracerTrail

File: lua/entities/ent_jack_barbedwire/init.lua [hunk 4]
```diff
@@ -190,4 +190,10 @@ function ENT:Think()
 
     return true
 
+end
+
+function ENT:ACF_PreDamage()
+    -- barbed wire can't be broken by ACF
+    return false
+
 end
```
Comment (wrefgtzweve) [inline lua/entities/ent_jack_barbedwire/init.lua]: Add a empty newline to the end of this file

File: lua/entities/ent_jack_seamine/init.lua [hunk 5]
```diff
@@ -125,3 +135,9 @@ function ENT:Use( activator )
 
     JID.genericUseEffect( activator )
 end
+
+function ENT:ACF_PreDamage()
+    -- ball busters are immune
+    return false
+
```
Comment (wrefgtzweve) [inline lua/entities/ent_jack_seamine/init.lua]: Newline should be removed

File: lua/entities/ent_jack_turret_base/init.lua [hunk 6]
```diff
@@ -133,6 +136,8 @@ function ENT:Initialize()
     self:SetMoveType( MOVETYPE_VPHYSICS )
     self:SetSolid( SOLID_VPHYSICS )
     self:SetUseType( SIMPLE_USE )
+    -- allow npcs to target me
```
Comment (wrefgtzweve) [inline lua/entities/ent_jack_turret_base/init.lua]: i prefer if you left comments like these after the call
```lua
self:AddFlags( FL_OBJECT ) -- allow npcs to target the turret
```

File: lua/entities/ent_jack_turret_base/init.lua [hunk 7]
```diff
@@ -157,6 +162,24 @@ function ENT:Initialize()
     self:SetDTBool( 0, self.HasAmmoBox )
     self:SetDTInt( 3, 0 )
     self.IFFTags = {}
+
+    timer.Simple( 0, function()
+        if not IsValid( self ) then return end
+        -- used by "heavy" turrets
```
Comment (wrefgtzweve) [inline lua/entities/ent_jack_turret_base/init.lua]: Comment should be on the same line as the code it affects

File: lua/entities/ent_jack_turret_base/init.lua [hunk 8]
```diff
@@ -797,31 +832,17 @@ function ENT:FireShot()
...
-    if self:GetClass() ~= "ent_jack_turret_plinker" then
-        sound.Play( self.NearShotNoise, ShootPos, 75, self.ShootSoundPitch )
-        sound.Play( self.FarShotNoise, ShootPos + Vector( 0, 0, 1 ), 110, self.ShootSoundPitch - 10 )
-
-    end
-
-    if self.AmmoType == "7.62x51mm" or self.AmmoType == ".338 Lapua Magnum" then
-        sound.Play( self.NearShotNoise, ShootPos + Vector( 0, 0, 1 ), 75, self.ShootSoundPitch + 10 )
+    local filterAllPlayers = RecipientFilter()
+    filterAllPlayers:AddAllPlayers()
```
Comment (wrefgtzweve) [inline lua/entities/ent_jack_turret_base/init.lua]: Why do this when EmitSound by default respects PAS? networking it to all players makes no logical sense to me. If sounds didn't properly work before use :AddPVS

File: lua/entities/ent_jack_turret_base/init.lua [hunk 9]
```diff
@@ -797,31 +832,17 @@ function ENT:FireShot()
...
+    local rangeLevelOffset = self.MaxRange / 300
 
-    if self.AmmoType == ".338 Lapua Magnum" then
-        sound.Play( self.NearShotNoise, ShootPos + Vector( 0, 0, 3 ), 75, self.ShootSoundPitch + 10 )
-        sound.Play( self.FarShotNoise, ShootPos + Vector( 0, 0, 4 ), 100, self.ShootSoundPitch + 5 )
-    end
+    self:EmitSound( self.NearShotNoise, 70 + rangeLevelOffset, self.ShootSoundPitch, 1, CHAN_WEAPON )
+    -- semi-global sound so players know where they're getting shot from
+    self:EmitSound( self.FarShotNoise, 90 + rangeLevelOffset, self.ShootSoundPitch - 10, 1, CHAN_WEAPON, 0, 0, filterAllPlayers )
+    -- play extra sounds to make it feel punchier.
```
Comment (wrefgtzweve) [inline lua/entities/ent_jack_turret_base/init.lua]: Comments should be on the same line as affected code

File: lua/entities/ent_jack_turret_grenade/init.lua [hunk 10]
```diff
@@ -136,8 +135,13 @@ function ENT:FireShot()
             return
         end
 
-        local distPow = Dist ^ 1.06
-        TargPos = TargPos + Vector( 0, 0, distPow / 60 )
+        if Dist <= 1000 and CurrTarg:IsPlayer() or CurrTarg:IsNPC() then
+            TargPos = CurrTarg:GetPos() -- aim at feet if they're close
+
```
Comment (wrefgtzweve) [inline lua/entities/ent_jack_turret_grenade/init.lua]: Newline can be removed

**Review by wrefgtzweve — COMMENTED** [review]
Mostly style but here and there ive had some questions about the code.

File: materials/entities/ent_jack_aidcomm.png
Comment (StrawWagen) [reply materials/entities/ent_jack_aidcomm.png:1]: Done, removed ents too

File: lua/entities/ent_jack_40mmgrenade/init.lua (hunk 3 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_40mmgrenade/init.lua]: 👍

File: lua/entities/ent_jack_barbedwire/init.lua (hunk 4 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_barbedwire/init.lua]: 👍

File: lua/entities/ent_jack_seamine/init.lua (hunk 5 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_seamine/init.lua]: 👍

File: lua/entities/ent_jack_turret_base/init.lua (hunk 6 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_base/init.lua]: 👍

File: lua/entities/ent_jack_turret_base/init.lua (hunk 7 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_base/init.lua]: 👍

File: lua/entities/ent_jack_turret_base/init.lua (hunk 8 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_base/init.lua]: PAS breaks alot with high soundlevel, seems to just not get to players on bigcity sometimes, etc.
Added this so players can always hear where they're getting shot from, no confusion
Changed to PVS 👍

File: lua/entities/ent_jack_turret_base/init.lua (hunk 9 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_base/init.lua]: 👍

File: lua/entities/ent_jack_turret_grenade/init.lua (hunk 10 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_grenade/init.lua]: 👍

**Review by wrefgtzweve — APPROVED** [review]

**Comment by wrefgtzweve** [comment]
I don't know if you should remove the entity files for the unused ents, if you think we'll never use them feel free to. Keeping them for future use however is no a issue, my initial remove comment was just about not having materials/content/downloadsize for unused stuff.

**Comment by StrawWagen** [comment]
Only extra ents left were the missile turrets, would've been retreading old ground gameplay wise, removin em is no issue

**Comment by wrefgtzweve** [comment]
👍

**Pull request merged**